### ISSUE TITLE
[Feature request] Optimize autograd/ATen when a gradient is clearly zero

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#4952]
When an input parameter ***clearly*** has zero gradient (not due to numerical coincidence), the current ATen guideline requires the backward function to return a zero tensor. Since there is no special treatment of zero tensors, autograd engine will still trace back the entire graph supporting that variable. This can happen easily when ATen functions have multiple outputs (e.g. many double backward functions). For example, 

```python
i = deep_nn_1(...)
w = deep_nn_2(...)
o = conv(i, w)
loss = autograd.grad(o.sum(), i, create_graph=True).sum()
loss.backward()
```

In conv double backward, `gI` only depends on `gO` and `ggW`, and `gW` only depends on `gO` and `ggI`. In this case, `ggW` is zero when doing the double backward, which means that `gI` should be zero and `deep_nn_2` doesn't need to be traversed at all. However, ATen conv double backward still [would output a zero `gI`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L686-L688) even if it already has the check on `ggW.defined()` to compute `gI`. Then the autograd engine will go through `deep_nn_2` unnecessarily.

**Comment by SsnL** [comment]
In general, giving zero tensors special treatment can also speed up operations including add and mul.

**Comment by zou3519** [comment]
What if we used a sparse zero-filled tensor to represent the gradient?

**Comment by albanD** [comment]
Note that we made a big progress towards this in https://github.com/pytorch/pytorch/pull/39400 and continuing it in https://github.com/pytorch/pytorch/pull/41490 and https://github.com/pytorch/pytorch/pull/41821.
